configure.ac

```diff
@@ -1715,72 +1704,78 @@ if test "x${HARDENING}" != "xtry"; then
-    CFLAGS_HARDENED_LIB=
-    LDFLAGS_HARDENED_EXE=
-    LDFLAGS_HARDENED_LIB=
-    relro=0
-    pie=0
-    bindnow=0
-    # daemons incl. libs: partial RELRO
-    flag="-Wl,-z,relro"
-    CC_CHECK_LDFLAGS(["${flag}"],
-                     [LDFLAGS_HARDENED_EXE="${LDFLAGS_HARDENED_EXE} ${flag}";
-                      LDFLAGS_HARDENED_LIB="${LDFLAGS_HARDENED_LIB} ${flag}";
-                      relro=1])
-    # daemons: PIE for both CFLAGS and LDFLAGS
-    if cc_supports_flag -fPIE; then
-        flag="-pie"
-        CC_CHECK_LDFLAGS(["${flag}"],
-                         [CFLAGS_HARDENED_EXE="${CFLAGS_HARDENED_EXE} -fPIE";
-                          LDFLAGS_HARDENED_EXE="${LDFLAGS_HARDENED_EXE} ${flag}";
-                          pie=1])
-    fi
-    # daemons incl. libs: full RELRO if sensible + as-needed linking
-    #                     so as to possibly mitigate startup performance
-    #                     hit caused by excessive linking with unneeded
-    #                     libraries
-    if test "${relro}" = 1 && test "${pie}" = 1; then
-        flag="-Wl,-z,now"
-        CC_CHECK_LDFLAGS(["${flag}"],
-                         [LDFLAGS_HARDENED_EXE="${LDFLAGS_HARDENED_EXE} ${flag}";
-                          LDFLAGS_HARDENED_LIB="${LDFLAGS_HARDENED_LIB} ${flag}";
-                          bindnow=1])
-    fi
-    if test "${bindnow}" = 1; then
-        flag="-Wl,--as-needed"
-        CC_CHECK_LDFLAGS(["${flag}"],
-                         [LDFLAGS_HARDENED_EXE="${LDFLAGS_HARDENED_EXE} ${flag}";
-                          LDFLAGS_HARDENED_LIB="${LDFLAGS_HARDENED_LIB} ${flag}"])
-    fi
-    # universal: prefer strong > all > default stack protector if possible
-    flag=
-    if cc_supports_flag -fstack-protector-strong; then
-        flag="-fstack-protector-strong"
-    elif cc_supports_flag -fstack-protector-all; then
-        flag="-fstack-protector-all"
-    elif cc_supports_flag -fstack-protector; then
-        flag="-fstack-protector"
-    fi
-    if test -n "${flag}"; then
-        CC_EXTRAS="${CC_EXTRAS} ${flag}"
-        stackprot=1
-    fi
-    if test "${relro}" = 1 \
-    || test "${pie}" = 1 \
-    || test "${stackprot}" = 1; then
-        AC_MSG_NOTICE([Hardening: relro=${relro} pie=${pie} bindnow=${bindnow} stackprot=${flag}])
-    else
-        AC_MSG_WARN([Hardening: no suitable features in the toolchain detected])
-    fi
-else
-    AC_MSG_NOTICE([Hardening: using custom flags])
-fi
+AS_IF([test "x${HARDENING}" = "xno"],
+      [AC_MSG_NOTICE([Hardening: explicitly disabled])],
+      [test "x${HARDENING}" = "xyes" || test "$(env | grep -Ec '^(C|LD)FLAGS_HARDENED_(EXE|LIB)=.')" = 0],
+      [
+          dnl We'll figure out on our own...
+          CFLAGS_HARDENED_EXE=
+          CFLAGS_HARDENED_LIB=
+          LDFLAGS_HARDENED_EXE=
+          LDFLAGS_HARDENED_LIB=
+          relro=0
+          pie=0
+          bindnow=0
+          # daemons incl. libs: partial RELRO
+          flag="-Wl,-z,relro"
```

Comment:
Mixing shell and autoconf comments here seems a little disorganized.  I know it's just from reorganizing the existing code.  It'd still be nice to do one or the other.

Reply:
They actually serve two different purposes -- dnl comments are strictly in configure.ac, # comments are copied to the generated configure script as well. I don't think anyone's adding configure to their summer reading list so I don't think it matters in practice.